Feature/csharp 8

jared-improbable opened the pull request:
Use more modern language features like `switch` expressions, `using` statements and deconstruction
Code quality improvements

jared-improbable committed:
Omit fields that reference their containing type
This should have never worked...
It only seems to be detected by the compiler in C# 8.0

jared-improbable committed:
Omit fields that reference their containing type
This should have never worked...
It only seems to be detected by the compiler in C# 8.0

zeroZshadow commented on Improbable/Postgres/Improbable.Postgres.CSharpCodeGen/ComponentGenerator.cs:
```diff
@@ -121,32 +117,30 @@ private static string PostgresTypeConversion(FieldDefinition f)
             {
                 case FieldType.Option:
-                    switch (f.OptionType.InnerType.ValueTypeSelector)
+                    if (f.OptionType.InnerType.ValueTypeSelector == ValueType.Primitive)
```
I think you can use if switch `when` here to remove the outmost if

jared-improbable replied:
Ah yes, nice!

zeroZshadow commented on Improbable/Postgres/Improbable.Postgres/PostgresOptions.cs:
```diff
@@ -51,34 +51,17 @@ public PostgresOptions(IPostgresOptions options)
+                PasswordFlagName => options.PostgresPassword,
+                DatabaseFlagName => options.PostgresDatabase,
+                AdditionalFlagName => options.PostgresAdditionalOptions,
+                _ => throw new InvalidOperationException($"Unknown Postgres flag {key}")
```
This should be a `ArgumentException` instead

zeroZshadow commented on Improbable/Stdlib/Improbable.Stdlib/WhenAllComponents.cs:
```diff
@@ -55,7 +55,7 @@ private void Added(AddComponentOp op)
             }
 
-            if (!counts.TryGetValue(op.EntityId, out var count))
+            if (!counts.TryGetValue(op.EntityId, out _))
```
```suggestion
            if (!counts.Contains(op.EntityId))
```

zeroZshadow replied:
This, or revert this change and set `count = 0;` inside the if, and counts.Add that. Then use `count + 1` over  `counts[op.EntityId] + 1` below.

zeroZshadow commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -178,61 +174,58 @@ public void StopSendingMetrics()
         }
 
-        private static List<LoginTokenDetails> GetDevelopmentLoginTokens(string host, ushort port, bool useInsecureConnection, string workerType, string pit)
+        private static IEnumerable<LoginTokenDetails> GetDevelopmentLoginTokens(string host, ushort port, bool useInsecureConnection, string workerType, string pit)
```
Might be a personal preference, but I would always return the highest type here (List<> in this case).
If a for each is done over the result it will use the List.Enumerator, while with the IEnumerable it wont.
Just a nit :P

jamiebrynes7 replied:
This is a interesting one. [Microsoft's guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections) suggest _not_ returning `List<T>`.

Rather they seem to favor using [`Collection<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1?view=netframework-4.8) for [a couple of reasons](https://blogs.msdn.microsoft.com/codeanalysis/2006/04/27/faq-why-does-donotexposegenericlists-recommend-that-i-expose-collectiont-instead-of-listt-david-kean/). Namely that:

* `Collection<>` has support for extensibility
* `List<>` exposes properties/methods that wouldn't typically be applicable. 

I don't think the first reason isn't a concern for this. And the second seems iffy to me as well (optionality is a bad thing?).

I think I agree with @zeroZshadow on this particular method, since we actually receive the underlying data as a `List<>` and `IEnumerable<>` forces you to allocate if you ever want to do something _other than_ forward-iterate over the return value. 

I can't think of a benefit to returning an `IEnumerable<>` in this case (there's no deferred execution benefits here that can come with `IEnumerable<>`) since its trivial for the caller to convert the `List<>` into the `IEnumerable<>` if they want.

jared-improbable replied:
Given the intended use case of this method, I don't think allocations are really going to be an overriding decision factor. I'm going to stick with `IEnumerable<>`

zeroZshadow commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -178,61 +174,58 @@ public void StopSendingMetrics()
-                return value.Value.LoginTokens;
+            if (value.Value.LoginTokens.Count == 0)
+            {
+                throw new Exception("No deployment returned for this project.");
```
Nit
I'm not sure if this should be an exception. Its totally expected to have no deployments running.

jared-improbable replied:
Agreed, I've removed this, it's up to client code to decide what to do

jamiebrynes7 commented on Improbable/Improbable.CSharpCodeGen/Types.cs:
```diff
@@ -153,436 +155,239 @@ public static class Types
         public static string GetFieldGetMethod(FieldDefinition field)
         {
-            switch (field.TypeSelector)
+            return field.TypeSelector switch
```
I :heart: pattern matching

jamiebrynes7 commented on Improbable/Schema/Improbable.Schema.Bundle/FieldDefinitionExtensions.cs:
```diff
@@ -0,0 +1,296 @@
+        {
+            if (!field.IsSingular())
+            {
+                throw new InvalidOperationException("Field type is not singular");
```
These exceptions should try to provide as much detail as possible for the field that caused the errors.

jared-improbable replied:
These are all going to indicate logic errors in a code generator rather than user input from schema errors, but I'll see what I can do.

jamiebrynes7 replied:
The extra information helps when debugging potential logic errors though

jared-improbable replied:
Looking at it more, I don't think there's much more to add other than the type it actually is, which I'll do.
You have the callstack, and you've messed up your type matching and are calling `IsSingular`, etc on a field that's not. It's really all in your code.

jamiebrynes7 replied:
All info is useful 😄

jamiebrynes7 commented on Improbable/Schema/Improbable.Schema.Bundle/FieldDefinitionExtensions.cs:
```diff
@@ -0,0 +1,296 @@
+                FieldType.List => field.ListType.InnerType.ValueTypeSelector == ValueType.Primitive,
+                FieldType.Map => throw new InvalidOperationException(),
+                FieldType.Singular => field.SingularType.Type.ValueTypeSelector == ValueType.Primitive,
+                _ => throw new ArgumentOutOfRangeException()
```
What's the rationale for checking the inner types for option & list, but not map? This seems a little odd to me.

jared-improbable replied:
I...don't know why I left that out. Will add it in.

jared-improbable replied:
Ah yes, I know, it's kind of nonsensical since maps have two types. I've added a message to the exception to check the Key/Value `TypeReference`s of maps individually.

jared-improbable committed:
Refactored ops processing so that they are deserialized in a backgrou... (#18)

* Refactored ops processing so that they are deserialized in a background task
Command results are now delivered from this background task
This makes it easier for users to avoid deadlocking themselves while bootstrapping their workers
e.g. Sending an Entity query to find service entities before starting the main loop

jared-improbable committed:
Bad merge

jared-improbable committed:
Merge branch 'feature/csharp-8' of github.com:spatialos/csharp-worker-template into feature/csharp-8

jared-improbable committed:
Extracted the specific casing/interpolation operations for names, types, commands, etc to simplify the code and make it more readable
Removed and collapsed various type switches and helpers

jared-improbable committed:
Compile fix

zeroZshadow commented on Improbable/Improbable.CSharpCodeGen/Case.cs:
```diff
@@ -27,53 +24,33 @@ public static string ToPascalCase(IEnumerable<string> parts)
                 .Select(SnakeCaseToPascalCase));
         }
-
+        
```
nit: whitespace

jared-improbable replied:
Fixed

zeroZshadow commented on Improbable/Postgres/Improbable.Postgres.CSharpCodeGen/ComponentGenerator.cs:
```diff
@@ -93,71 +93,41 @@ public static string InitializeDatabase(string tableName)
-            return string.Empty;
+                FieldType.Option when f.HasPrimitive(PrimitiveType.String) => "::text",
+                FieldType.Option when f.HasPrimitive() => string.Empty,
+                FieldType.Option => throw new InvalidOperationException($"Unsupported schema type '{f.OptionType.InnerType.Type}'. Only primitive types can be converted from Postgres."),
```
This can probably be combined with the same error below using a fallthrough?

jared-improbable replied:
Ooops, it actually can't because it's Option vs Singular
You found a bug :)

zeroZshadow commented on Improbable/Schema/Improbable.Schema.Bundle/TypeDescription.cs:
```diff
@@ -42,7 +42,8 @@ public TypeDescription(string qualifiedName, Bundle bundle)
             IsRestricted = qualifiedName.StartsWith("improbable.restricted");
 
-            NestedEnums = bundle.Enums.Where(e => e.Value.OuterType == qualifiedName).Select(type => bundle.Enums[type.Key]).ToList();
+            NestedEnums = bundle.Enums.Where(e => e.Value.OuterType == qualifiedName)
```
Might just be my preference, but I usually put every part of the LINQ statement on a new line.

zeroZshadow commented on Improbable/Schema/Improbable.Schema.Bundle/TypeDescription.cs:
```diff
@@ -103,7 +94,13 @@ public TypeDescription(string qualifiedName, Bundle bundle)
                 }
 
-                return allowed;
+                var allowed2 = !IsFieldTypeRecursive(bundle, qualifiedName, f);
```
This variable name is not oke

jared-improbable replied:
Fixed

zeroZshadow commented on Improbable/Stdlib/Improbable.Stdlib/ComponentCollection.cs:
```diff
@@ -135,14 +133,15 @@ public bool Contains(EntityId entityId)
+            freeSlots.Enqueue(index);
+            lookup.Remove(entityId);
+            return true;
+
```
nit: newline

jared-improbable replied:
Fixed

jamiebrynes7 commented on Improbable/Improbable.CSharpCodeGen/Case.cs:
```diff
@@ -27,53 +24,33 @@ public static string ToPascalCase(IEnumerable<string> parts)
-
-        public static string AllCapsSnakeCaseToPascalCase(string screamingSnake)
-        {
-            return screamingSnake.Split(new[] { "_" }, StringSplitOptions.RemoveEmptyEntries)
```
🐍 😱

jamiebrynes7 commented on Improbable/Schema/Improbable.Schema.Bundle/FieldDefinitionExtensions.cs:
```diff
@@ -0,0 +1,122 @@
+            };
+        }
+
+        public static bool CanPrimitiveBeNull(this FieldDefinition field)
```
Can `Entity` be null? Or are you not supporting that, I can't recall

jared-improbable replied:
Ah nope, not supported

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -74,17 +115,20 @@ public static Task<WorkerConnection> ConnectAsync(IWorkerOptions workerOptions,
-                return new WorkerConnection(connection);
+            if (connection.GetConnectionStatusCode() != ConnectionStatusCode.Success)
+            {
+                throw new Exception($"{connection.GetConnectionStatusCode()}: {connection.GetConnectionStatusCodeDetailString()}");
```
Duplicate block here

jared-improbable replied:
Weird, fixed!

jamiebrynes7 commented on Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:
```diff
@@ -107,18 +151,21 @@ public static async Task<WorkerConnection> ConnectAsync(ILocatorOptions options,
-                return new WorkerConnection(connection);
+            if (connection != null && connection.GetConnectionStatusCode() != ConnectionStatusCode.Success)
+            {
+                throw new Exception($"{connection.GetConnectionStatusCode()}: {connection.GetConnectionStatusCodeDetailString()}");
```
And here

jared-improbable replied:
And fixed

jamiebrynes7 reviewed (approved):
LGTM % nits

jared-improbable committed:
CR Feedback
